### PR TITLE
vitolscn code review 2

### PR DESCRIPTION
**Pull request by vitolscolin**
Analysis of the program:
Program is created to keep track of the current price of select stocks. Stocks that are tracked can be selected by the user from a JSON string. 

Was the program available in UC Github on time?
Yes, no issues there. 

Is the program documented/commented well enough for you to understand?:
Yes, plenty of documentation to allow me to understand the program.

Does the program compile?
Compiled without issues. 

Rationale behind your changes.
I looked at best practices, along with combing through the program, to try and remove any unneeded features. 

Three technical concepts I learned:
I learned a bit more about firebase's drop-in login page. More specially, the reasoning behind the Auth_Code and Response_Code variables. This will help me greatly as I am focusing on the login page.
I learned more about how JSON files are read and how to use retrofit clients to help parse those JSON files.
Learned the difference between log.e and log.d and best practices behind log files.

### CODE REVIEW

File: app/src/main/java/edu/uc/coffeens/stocktracker/services/RetrofitClientInstance.kt [hunk 1]
```diff
@@ -8,15 +8,15 @@ import retrofit2.converter.gson.GsonConverterFactory
...
+    private var retrofit: Retrofit? = null
     private val BASE_URL = "https://gist.githubusercontent.com/"
 
     val retrofitInstance: Retrofit?
         get() {
             //Has this object been created?
             if (retrofit == null) {
                 // Create retrofit instance
-                retrofit = retrofit2.Retrofit.Builder()
+                retrofit = Retrofit.Builder()
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/services/RetrofitClientInstance.kt:19]: Optimized how the retrofit builder is built. Looks a bit cleaner, and easier to read.

File: app/src/main/java/edu/uc/coffeens/stocktracker/services/StockService.kt [hunk 2]
```diff
@@ -15,12 +16,12 @@ class StockService {
      */
     fun fetchStock(): MutableLiveData<ArrayList<Stock>> {
 
-        var _stock = MutableLiveData<ArrayList<Stock>>()
+        val _stock = MutableLiveData<ArrayList<Stock>>()
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/services/StockService.kt:19]: updated to val as the Array currently is a constant.

File: app/src/main/java/edu/uc/coffeens/stocktracker/services/StockService.kt [hunk 3]
```diff
@@ -15,12 +16,12 @@ class StockService {
...
         val service = RetrofitClientInstance.retrofitInstance?.create(IStockDAO::class.java)
         val call = service?.getAllStocks()
         call?.enqueue(object : Callback<ArrayList<Stock>> {
             override fun onFailure(call: Call<ArrayList<Stock>>, t: Throwable) {
-                print("Unable to retrieve service response.")
+                Log.e("error", "Unable to retrieve service response.")
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/services/StockService.kt:24]: Changed to log.e, this will cause the error to output in the error log.

File: app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt [hunk 4]
```diff
@@ -2,14 +2,11 @@ package edu.uc.coffeens.stocktracker.ui.main
...
 import android.content.Intent
 import android.os.Bundle
-import android.util.Log
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
 import android.widget.ArrayAdapter
-import android.widget.Button
 import androidx.fragment.app.Fragment
-import androidx.lifecycle.Observer
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt:12]: optimized imports.

File: app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt [hunk 5]
```diff
@@ -92,11 +89,5 @@ class MainFragment : Fragment() {
         }
     }
 
-    fun getClickedStock(view: View) {
-        Log.d("[SEARCH BAR]", "Clicked a stock.")
-    }
-
-
-
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt:100]: deleted function, at time of code review, the function was never called.

File: app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt [hunk 6]
```diff
@@ -69,7 +66,7 @@ class MainFragment : Fragment() {
      * Function to be called by the log on button
      */
     private fun logon() {
-        var providers = arrayListOf(
+        val providers = arrayListOf(
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt:69]: updated var to val as providers seems to be a constant.

File: app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainViewModel.kt [hunk 7]
```diff
@@ -36,10 +34,10 @@ class MainViewModel : ViewModel() {
             .document()
             .set(stocks)
             .addOnSuccessListener {
-                Log.d("Firebase", "document saved")
+                Log.e("Firebase", "document saved")
             }
             .addOnFailureListener {
-                Log.d("Firebase", "save failed")
+                Log.e("Firebase", "save failed")
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainViewModel.kt]: updated to log.e so if these errors occur, they can be spotted in something other than debug mode.

File: app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt [hunk 8]
```diff
@@ -92,11 +89,5 @@ class MainFragment : Fragment() {
         }
     }
 
-    fun getClickedStock(view: View) {
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt:95]: deleted function as function was not called at the time of code review.

File: app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt [hunk 9]
```diff
@@ -50,7 +47,7 @@ class MainFragment : Fragment() {
     override fun onActivityCreated(savedInstanceState: Bundle?) {
         super.onActivityCreated(savedInstanceState)
         viewModel = ViewModelProvider(this).get(MainViewModel::class.java)
-        viewModel.stock.observe(viewLifecycleOwner, Observer { stocks ->
+        viewModel.stock.observe(viewLifecycleOwner, { stocks ->
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt:50]: cleaned up how the observer was built, makes it a little cleaner.

File: app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt [hunk 10]
```diff
@@ -2,14 +2,11 @@ package edu.uc.coffeens.stocktracker.ui.main
 
 import android.content.Intent
 import android.os.Bundle
-import android.util.Log
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainFragment.kt:5]: optimized imports.

File: app/src/main/java/edu/uc/coffeens/stocktracker/services/RetrofitClientInstance.kt [hunk 11]
```diff
@@ -8,15 +8,15 @@ import retrofit2.converter.gson.GsonConverterFactory
  */
 object RetrofitClientInstance {
 
-    private var retrofit: Retrofit? = null;
```
Comment (vitolscolin) [inline app/src/main/java/edu/uc/coffeens/stocktracker/services/RetrofitClientInstance.kt:11]: unneeded semicolon

File: app/src/main/res/layout/main_fragment.xml [hunk 12]
```diff
@@ -11,7 +11,7 @@
         android:id="@+id/actStock"
         android:layout_width="0dp"
         android:layout_height="wrap_content"
-        android:hint="Enter stock"
+        android:hint="@string/actStock_Text"
```
Comment (vitolscolin) [inline app/src/main/res/layout/main_fragment.xml:14]: changed hard-coded "enter stock" in the text box to a value. This will help if the textbox wants to be duplicated to allow the user to track multiple stocks.

File: app/src/main/res/layout/main_fragment.xml [hunk 13]
```diff
@@ -21,7 +21,7 @@
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
         android:layout_marginBottom="52dp"
-        android:text="Login"
+        android:text="@string/btnLogin_Text"
```
Comment (vitolscolin) [inline app/src/main/res/layout/main_fragment.xml:24]: changed the hard-coded text "login" in the login button to a value. Not particularly needed, but it's best practice to avoid hard-coded text.

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -1,3 +1,5 @@
 <resources>
     <string name="app_name">StockTracker</string>
+    <string name="actStock_Text">Enter stock</string>
+    <string name="btnLogin_Text">Login</string>
```
Comment (vitolscolin) [inline app/src/main/res/values/strings.xml:4]: values were added to the strings.xml file.

File: app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainViewModel.kt [hunk 15]
```diff
@@ -30,16 +28,4 @@ class MainViewModel : ViewModel() {
...
-        firestore.collection("Stocks")
-            .document()
-            .set(stocks)
-            .addOnSuccessListener {
-                Log.d("Firebase", "document saved")
-            }
-            .addOnFailureListener {
-                Log.d("Firebase", "save failed")
-            }
-    }
```
Comment (discospiff) [inline app/src/main/java/edu/uc/coffeens/stocktracker/ui/main/MainViewModel.kt:44]: Why did you remove this?

**Comment by discospiff** [comment]
Several good concepts to harvest from this review.
- Use proper logging instead of print.
- var to val.
- String externalization.
- Remove unused code.

I'd like to verify why one function was removed (likely it's unused).  Then, the branch is worth merging.
